app/models/foreman_tasks/recurring_logic.rb

```diff
@@ -189,5 +193,9 @@ def self.cronline_hash(recurring_type, time_hash, days, days_of_week_hash)
                      end
       hash.select { |key, _| allowed_keys.include? key }
     end
+
+    def valid_purpose?
+      !(purpose.present? && self.class.where(:purpose => purpose, :state => %w[active disabled]).any?)
```

Comment:
I'm a bit worried about concurrent writes here, especially if we will do something like "create a new RL with given purpose if it does not already exist" from an initializer.

Reply:
Yes, concurrent cases will probably be tricky. Do we know about the places where they might occur?
We could go for ActiveRecord transactions or I can try to conjure a partial unique index for postgres, though I am not sure if that would help. Any ideas?

Reply:
Initializers will most likely be prime suspects. How about instead of the purpose being a property of a recurring logic itself, it would be a separate table where we could leverage unique index? The only downside is RLs would have to drop records from this table when they finish

Reply:
That might work, though we also need to retain information about finished RLs with a particular purpose - so we'd probably need another relationship for that.

I looked into the partial index which seems like a less involved approach and pushed a change, let me know what you think.

Reply:
Seems to work well :+1: